xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+      }
+      // TODO(chengyuanzhang): send CDS/EDS requests if CDS/EDS watcher presents.
+    }
+  }
+
+  private final class AdsStream implements StreamObserver<DiscoveryResponse> {
+    private final AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub;
+
+    private StreamObserver<DiscoveryRequest> requestWriter;
+    private boolean responseReceived;
+    private boolean closed;
+
+    // Last successfully applied version_info for each resource type. Starts with empty string.
+    // A version_info is used to update management server with client's most recent knowledge of
+    // resources.
+    private String ldsVersion = "";
+    private String rdsVersion = "";
+
+    // Response nonce for the most recently received discovery responses of each resource type.
+    // Client initiated requests start response nonce with empty string.
+    // A nonce is used to indicate the specific DiscoveryResponse each DiscoveryRequest
+    // corresponds to.
+    // A nonce becomes stale following a newer nonce being presented to the client in a
+    // DiscoveryResponse.
+    private String ldsRespNonce = "";
+    private String rdsRespNonce = "";
+
+    // Most recently requested resource name(s) for each resource type. Note the resource_name in
+    // LDS requests will always be "xds:" URI (including port suffix if present).
+    @Nullable
+    private String rdsResourceName;
+
+    private AdsStream(AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub) {
+      this.stub = checkNotNull(stub, "stub");
+    }
+
+    private void start() {
+      requestWriter = stub.withWaitForReady().streamAggregatedResources(this);
+    }
+
+    @Override
+    public void onNext(final DiscoveryResponse response) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          responseReceived = true;
+          String typeUrl = response.getTypeUrl();
+          if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+            handleLdsResponse(response);
+          } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+            handleRdsResponse(response);
+          }
+          // TODO(zdapeng): add CDS/EDS response handles.
+        }
+      });
+    }
+
+    @Override
+    public void onError(final Throwable t) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.fromThrowable(t).augmentDescription("ADS stream [" + this + "] had an error"));
+        }
+      });
+    }
+
+    @Override
+    public void onCompleted() {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.UNAVAILABLE.withDescription("ADS stream [" + this + "] was closed"));
+        }
+      });
+    }
+
+    private void handleStreamClosed(Status error) {
+      checkArgument(!error.isOk(), "unexpected OK status");
```

Comment:
In both the calls to `handleStreamClosed` you are passing non-Ok status so this check will always fail and throw an exception. So the rest of the code in this method is never exercised - is that correct?

Reply:
Yes, that's the expected behavior be all gRPC server implementations. A streaming server should not set status to `OK` if it terminates the stream. You can see [`grpclb`](https://github.com/grpc/grpc-java/blob/b05ce13df2307de40ee84a0476954194bdde89e2/grpclb/src/main/java/io/grpc/grpclb/GrpclbState.java#L635) is also doing the same thing.